private/buf/cmd/buf/command/curl/curl.go

```diff
@@ -0,0 +1,832 @@
+	insecureFlagName   = "insecure"
+
+	// Timeout flags
+	noKeepAliveFlagName    = "no-keepalive"
+	keepAliveFlagName      = "keepalive-time"
+	connectTimeoutFlagName = "connect-timeout"
+
+	// Header and request body flags
+	userAgentFlagName = "user-agent"
+	headerFlagName    = "header"
+	dataFlagName      = "data"
+	outputFlagName    = "output"
+)
+
+const (
+	protocolConnect = "connect"
+	protocolGRPC    = "grpc"
+	protocolGRPCWeb = "grpcweb"
+
+	reflectVersionAuto    = "auto"
+	reflectVersionV1      = "v1"
+	reflectVersionV1Alpha = "v1alpha"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <url>",
+		Short: "Invoke an RPC endpoint, a la 'cURL'.",
+		Long: `This command helps you invoke HTTP RPC endpoints on a server that uses gRPC
+or Connect.
+
+By default, server reflection is used, unless the --reflect flag is set to
+false. Without server reflection, a --schema flag must be provided to indicate
+the Protobuf schema for the method being invoked.
+
+The only positional argument is the URL of the RPC method to invoke. The name of
+the method to invoke comes from the last two path components of the URL, which
+should be the fully-qualified service name and method name, respectively.
+
+The input request is specified via the -d or --data flag. If absent, an empty
+request is sent. If the flag value starts with an at-sign (@), then the rest of
+the flag value is interpreted as a filename from which to read the request body.
+If that filename is just a dash (-), then the request body is read from stdin.
+The request body is a JSON document that contains the JSON formatted request
+message. If the RPC method being invoked is a client-streaming method, the
+request body may consist of multiple JSON values, appended to one another.
+Multiple JSON documents should usually be separated by whitespace, though this
+is not strictly required unless the request message type has a custom JSON
+representation that is not a JSON object.
+
+Request metadata (i.e. headers) are defined using -H or --header flags. The flag
+value is in "name: value" format. But if it starts with an at-sign (@), the rest
+of the value is interpreted as a filename from which headers are read, each on a
+separate line. If the filename is just a dash (-), then the headers are read
+from stdin.
+
+If headers and the request body are both to be read from the same file (or both
+read from stdin), the file must include headers first, then a blank line, and
+then the request body.
+
+The URL can use either http or https as the scheme. If http is used then HTTP
+1.1 will be used unless the --http2-prior-knowledge flag is set. If https is
+used then HTTP/2 will be preferred during protocol negotiation and HTTP 1.1 used
+only if the server does not support HTTP/2.
+
+The default RPC protocol used will be gRPC. However this protocol cannot work
```

Comment:
why not default to connect?

Reply:
TBH, 99.9% of requests using the Connect protocol do not really need a `buf curl` tool: it was designed to be cURL-friendly and supports JSON out of the box. The main value of this seemed to be the fact that it can get the schema and then translate human-friendlier JSON to binary protobuf, which is mainly needed for grpc and grpcweb servers.

Though it's true that it does also provide a much easier way to do streaming than with cURL (which would require some shenanigans to wrap each JSON value in an envelope).

I also worried that Connect doesn't have enough adoption to be the default. I think most people making use of this feature would be targeting gRPC servers. I wonder what @akshayjshah thinks.

Reply:
I've gone back and forth on this, but I'd probably either (1) have no default or (2) default to connect, although I understand your motivations (and don't think they're off-base). happy to touch base briefly offline

Reply:
Changed default to connect